backends/events/sdl/sdl-events.cpp

```diff
@@ -409,9 +415,329 @@ Common::KeyCode SdlEventSource::SDLToOSystemKeycode(const SDL_Keycode key) {
+					if ((xrel * xrel + yrel * yrel) < maxRSquared) {
+						if (numFingersDown == 2 || numFingersDown == 1) {
+							Uint8 simulatedButton = 0;
+							if (numFingersDown == 2) {
+								simulatedButton = SDL_BUTTON_RIGHT;
+								// need to raise the button later
+								_touchPanels[port]._simulatedClickStartTime[1] = event->tfinger.timestamp;
+							} else if (numFingersDown == 1) {
+								simulatedButton = SDL_BUTTON_LEFT;
+								// need to raise the button later
+								_touchPanels[port]._simulatedClickStartTime[0] = event->tfinger.timestamp;
+								if (!isTouchPortTouchpadMode(port)) {
+									convertTouchXYToGameXY(event->tfinger.x, event->tfinger.y, &x, &y);
+								}
+							}
+
+							event->type = SDL_MOUSEBUTTONDOWN;
+							event->button.button = simulatedButton;
+							event->button.x = x;
+							event->button.y = y;
+						}
+					}
+				}
+			} else if (numFingersDown == 1) {
+				// when dragging, and the last finger is lifted, the drag is over
+				if (!isTouchPortTouchpadMode(port)) {
+					convertTouchXYToGameXY(event->tfinger.x, event->tfinger.y, &x, &y);
+				}
+				Uint8 simulatedButton = 0;
+				if (_touchPanels[port]._multiFingerDragging == DRAG_THREE_FINGER)
+					simulatedButton = SDL_BUTTON_RIGHT;
+				else {
+					simulatedButton = SDL_BUTTON_LEFT;
+				}
+				event->type = SDL_MOUSEBUTTONUP;
+				event->button.button = simulatedButton;
+				event->button.x = x;
+				event->button.y = y;
+				_touchPanels[port]._multiFingerDragging = DRAG_NONE;
+			}
+		}
+	}
+}
+
+void SdlEventSource::preprocessFingerMotion(SDL_Event *event) {
+	// front (0) or back (1) panel
+	SDL_TouchID port = event->tfinger.touchId;
+	// id (for multitouch)
+	SDL_FingerID id = event->tfinger.fingerId;
+
+	// find out how many fingers were down before this event
+	int numFingersDown = 0;
+	for (int i = 0; i < MAX_NUM_FINGERS; i++) {
+		if (_touchPanels[port]._finger[i].id >= 0) {
+			numFingersDown++;
+		}
+	}
+
+	if (numFingersDown >= 1) {
+		int x = _mouseX;
+		int y = _mouseY;
+		int xMax = _graphicsManager->getWindowWidth() - 1;
+		int yMax = _graphicsManager->getWindowHeight() - 1;
+
+		if (!isTouchPortTouchpadMode(port)) {
+			convertTouchXYToGameXY(event->tfinger.x, event->tfinger.y, &x, &y);
+		}	else {
+			// for relative mode, use the pointer speed setting
+			float speedFactor = 1.0;
+
+			switch (ConfMan.getInt("kbdmouse_speed")) {
+			// 0.25 keyboard pointer speed
+			case 0:
+				speedFactor = 0.25;
+				break;
+			// 0.5 speed
+			case 1:
+				speedFactor = 0.5;
+				break;
+			// 0.75 speed
+			case 2:
+				speedFactor = 0.75;
+				break;
+			// 1.0 speed
+			case 3:
+				speedFactor = 1.0;
+				break;
+			// 1.25 speed
+			case 4:
+				speedFactor = 1.25;
+				break;
+			// 1.5 speed
+			case 5:
+				speedFactor = 1.5;
+				break;
+			// 1.75 speed
+			case 6:
+				speedFactor = 1.75;
+				break;
+			// 2.0 speed
+			case 7:
+				speedFactor = 2.0;
+				break;
+			default:
+				speedFactor = 1.0;
+			}
```

Comment:
This could be simplified as follows:

```suggestion
			const int32 kbdMouseSpeed = CLIP<int32>(ConfMan.getInt("kbdmouse_speed"), 0, 7);
			float speedFactor = (kbdMouseSpeed  + 1) * 0.25;
```

Reply:
True, but it's not a new code, I just move the code around. I suggest this change to be a separate one